pandas/core/frame.py

```diff
@@ -6237,13 +6237,30 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
         if not self.columns.is_unique:
             raise ValueError("columns must be unique")
 
-        result = self[column].explode()
-        return (
+        named_index = any(i for i in self.index.names)
+
+        column_with_index = self[column].reset_index()
+
+        result = (
             self.drop([column], axis=1)
-            .join(result)
-            .reindex(columns=self.columns, copy=False)
+            .reset_index()
+            .join(column_with_index[column].explode())
         )
 
+        if isinstance(self.index, MultiIndex):
+            result.index = pandas.MultiIndex.from_frame(
+                result.iloc[:, : self.index.nlevels]
+            )
+        else:
+            result.index = result.iloc[:, 0]
+
+        if not named_index:
+            result.index.names = [None] * self.index.nlevels
```

Comment:
Didn't run this, but I have the feeling that if the names of the MultiIndex levels in the input DataFrame are something like `['foo', None, 'bar']`, in the output will be `['foo', 'level_0', 'bar']`. If that's the case, besides fixing it, would be nice to add an example.

Reply:
we don’t implicitly ever change names like this

Reply:
@datapythonista you're right. I didn't think partially named multiindices were a thing, sorry for not having checked. Fix and test cases added.

@jreback sorry about that - should it be `result.index.rename()` instead?

Thanks to both of you for your time.

By the way, even on the master branch (without any of my changes applied), this sometimes fails:
```
pytest pandas/tests/computation/test_eval.py -k test_binary_arith_ops
```
I just ran it 10 times, and 6 of them had a failure.

From what I can discern from the Travis CI build logs, that's what's currently failing here